### PR TITLE
module update

### PR DESCRIPTION
**Pull request by JordanYates**
Update tflite-micro to a version compatible with CMSIS-NN v6.0.0

### CODE REVIEW

File: zephyr_static_export.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+make -C $MAKE_FOLDER TENSORFLOW_ROOT=`pwd`/
+
+# Copy the files we need
+mkdir -p third_party_static/flatbuffers
+mkdir -p third_party_static/gemmlowp
+mkdir -p third_party_static/ruy/ruy
+mkdir -p third_party_static/kissfft
+
+cp $DL_FOLDER/flatbuffers/LICENSE third_party_static/flatbuffers/LICENSE
+cp -r $DL_FOLDER/flatbuffers/include third_party_static/flatbuffers/include
```
Comment (ithinuel) [inline zephyr_static_export.sh:16]: Makes it go in `third_party_static/flatbuffers/include/include` (duplicated include)

```suggestion
cp -r $DL_FOLDER/flatbuffers/include third_party_static/flatbuffers
```

File: zephyr_static_export.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+cp $DL_FOLDER/gemmlowp/LICENSE third_party_static/gemmlowp/LICENSE
+cp -r $DL_FOLDER/gemmlowp/fixedpoint third_party_static/gemmlowp/fixedpoint
+cp -r $DL_FOLDER/gemmlowp/internal third_party_static/gemmlowp/internal
```
Comment (ithinuel) [inline zephyr_static_export.sh:20]:
```suggestion
cp -r $DL_FOLDER/gemmlowp/fixedpoint third_party_static/gemmlowp
cp -r $DL_FOLDER/gemmlowp/internal third_party_static/gemmlowp
```

File: zephyr_static_export.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+cp $DL_FOLDER/ruy/LICENSE third_party_static/ruy/LICENSE
+cp -r $DL_FOLDER/ruy/ruy/profiler third_party_static/ruy/ruy/profiler
+
+cp $DL_FOLDER/kissfft/COPYING third_party_static/kissfft/COPYING
+cp $DL_FOLDER/kissfft/_kiss_fft_guts.h third_party_static/kissfft/_kiss_fft_guts.h
+cp $DL_FOLDER/kissfft/kiss_fft.c third_party_static/kissfft/kiss_fft.c
+cp $DL_FOLDER/kissfft/kiss_fft.h  third_party_static/kissfft/kiss_fft.h
+cp $DL_FOLDER/kissfft/kissfft.hh third_party_static/kissfft/kissfft.hh
+cp -r $DL_FOLDER/kissfft/tools third_party_static/kissfft/tools
```
Comment (ithinuel) [inline zephyr_static_export.sh:30]:
```suggestion
cp -r $DL_FOLDER/kissfft/tools third_party_static/kissfft
```

File: zephyr_static_export.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+cp $DL_FOLDER/flatbuffers/LICENSE third_party_static/flatbuffers/LICENSE
+cp -r $DL_FOLDER/flatbuffers/include third_party_static/flatbuffers/include
+
+cp $DL_FOLDER/gemmlowp/LICENSE third_party_static/gemmlowp/LICENSE
+cp -r $DL_FOLDER/gemmlowp/fixedpoint third_party_static/gemmlowp/fixedpoint
+cp -r $DL_FOLDER/gemmlowp/internal third_party_static/gemmlowp/internal
+
+cp $DL_FOLDER/ruy/LICENSE third_party_static/ruy/LICENSE
+cp -r $DL_FOLDER/ruy/ruy/profiler third_party_static/ruy/ruy/profiler
```
Comment (ithinuel) [inline zephyr_static_export.sh:23]:
```suggestion
cp -r $DL_FOLDER/ruy/ruy/profiler third_party_static/ruy/ruy
```

**Review by ithinuel — CHANGES_REQUESTED** [review]
This is based on the `zephyr` branch rather than the `main` branch.

The addition of the export script seems to be a good idea to simplify future updates however it yields the wrong paths.

To review/validate this PR I ran:

```
git fetch pull/6/HEAD:PR6
git checkout main
git merge 71fba59b
# add & run the script
# add third_party_static
# commit
git diff PR6
```

**Review by nashif — APPROVED** [review]

**Comment by nashif** [comment]
ok, this is a mess, no idea what is going on here and how we got here :(

**Comment by JordanYates** [comment]
> This is based on the `zephyr` branch rather than the `main` branch.
> 
> The addition of the export script seems to be a good idea to simplify future updates however it yields the wrong paths.
> 
> To review/validate this PR I ran:
> 
> ```
> git fetch pull/6/HEAD:PR6
> git checkout main
> git merge 71fba59b
> # add & run the script
> # add third_party_static
> # commit
> git diff PR6
> ```

Paths are fine when I run it and my simple reproduction behaves as I expect, no duplicated directories:
```
(.zephyr_venv) jordan@TAURUS:~$ mkdir tmp
(.zephyr_venv) jordan@TAURUS:~$ cp -r ~/code/zephyr/zephyr/include/ ./tmp/include
(.zephyr_venv) jordan@TAURUS:~$ ls -ls tmp/include/
total 4
4 drwxr-xr-x 57 jordan jordan 4096 Jun 27 20:18 zephyr
```

**Comment by JordanYates** [comment]
> ok, this is a mess, no idea what is going on here and how we got here :(

I agree, a giant mess and why I would prefer a branch :)

**Comment by JordanYates** [comment]
> This is based on the zephyr branch rather than the main branch.

Nope, this is based purely on the upstream tensorflow micro main branch.
```
(.zephyr_venv) jordan@TAURUS:~/code/zephyr/optional/modules/lib/tflite-micro$ git remote -v
jordan  git@github.com:JordanYates/tflite-micro.git (fetch)
jordan  git@github.com:JordanYates/tflite-micro.git (push)
origin  git@github.com:tensorflow/tflite-micro.git (fetch)
origin  git@github.com:tensorflow/tflite-micro.git (push)
upstream        https://github.com/zephyrproject-rtos/tflite-micro (fetch)
upstream        https://github.com/zephyrproject-rtos/tflite-micro (push)
(.zephyr_venv) jordan@TAURUS:~/code/zephyr/optional/modules/lib/tflite-micro$ git log --oneline
e1eecbee (HEAD, jordan/zephyr-v3.7, zephyr-v3.7, manifest-rev, 240613_export) third_party_static: exported
73258a3e zephyr_static_export.sh: added
8be43d15 zephyr: module.yml: added
71fba59b (origin/main) Fail per-channel quantized FullyConnected layers (#2602)
```
Or shown a different way: https://github.com/JordanYates/tflite-micro/commits/zephyr-v3.7/

**Comment by stephanosio** [comment]
> > ok, this is a mess, no idea what is going on here and how we got here :(
> 
> I agree, a giant mess and why I would prefer a branch :)

https://github.com/zephyrproject-rtos/zephyr/pull/73995#discussion_r1656909607

Let us create a new branch -- it makes sense and everyone keeps their sanity.

Now the only question is: `zephyr-v3.7` vs `vTFLITEMICROVER-zephyr` ...

**Comment by JordanYates** [comment]
> Now the only question is: `zephyr-v3.7` vs `vTFLITEMICROVER-zephyr` ...

The name is almost irrelevant since all that matters is the commit hash.
Someone with push rights make a unilateral decision and I will support it.

**Comment by stephanosio** [comment]
> The name is almost irrelevant since all that matters is the commit hash.

Branch names do matter because they reflect how upstream syncs are meant to be handled.

> Someone with push rights make a unilateral decision and I will support it.

I say we go and stick with `zephyr-v3.7` (i.e. `zephyr-vZEPHYRVER`) because, that way, it is easier to backport module patches for specific Zephyr release in the future -- this is especially important for an LTS too ...

@nashif Do you agree?

**Comment by ithinuel** [comment]
> > This is based on the zephyr branch rather than the main branch.
>
> Nope, this is based purely on the upstream tensorflow micro main branch.

Indeed, my bad.

> Paths are fine when I run it and my simple reproduction behaves as I expect, no duplicated directories:

No they are not.

This file is in [`third_party_static/flatbuffers/include/include/flatbuffers/base.h`](https://github.com/zephyrproject-rtos/tflite-micro/pull/6/files#diff-edce4d56aeb1e384b44875eae2ea5c233957d7c932a0d8f27e9b215a38504a93) and [`third_party_static/flatbuffers/include/flatbuffers/base.h`](https://github.com/zephyrproject-rtos/tflite-micro/pull/6/files#diff-fa78dd3bcd0693e8b6cad7c8e721380f9303ff9bfe171d4a08b4f3d235c53f4a) (and all the other files and directories in flatbuffers.

You can check around `third_party_static/gemmlowp/fixedpoint/fixedpoint.h` as there are less files and you can see both the outer and the inner replicated directories.

Screenshot 2024-06-27 at 12 02 55

**Comment by nashif** [comment]
I prefer to use zephyr, when someone lands on the pages, they need to see the code being used and not some old commits.
We can tag and archive existing state and reset the 'zephyr' branch with the code.

**Comment by nashif** [comment]
Just give me a sha that can be used and I will  do that directly.

**Comment by stephanosio** [comment]
> I prefer to use zephyr, when someone lands on the pages, they need to see the code being used and not some old commits. We can tag and archive existing state and reset the 'zephyr' branch with the code.

@nashif So how are we supposed to handle tflite-micro backport patches for Zephyr v3.7 when Zephyr is at v4.5 and the tflite-module `zephyr` branch is well into the future?

It is much easier to open/review a "backport patch PR" to `zephyr-v3.7` branch than to do some behind-the-scenes chat and tagging magic.

**Comment by JordanYates** [comment]
> You can check around third_party_static/gemmlowp/fixedpoint/fixedpoint.h as there are less files and you can see both the outer and the inner replicated directories.

Turns out we were both right. You were correct that files were duplicated, they were accidentally committed while testing the export script.

Reverted the entire export commit and re-ran the script, files now match the current structure.

**Comment by nashif** [comment]
> @nashif So how are we supposed to handle tflite-micro backport patches for Zephyr v3.7 when Zephyr is at v4.5 and the tflite-module `zephyr` branch is well into the future?
> 
> It is much easier to open/review a "backport patch PR" to `zephyr-v3.7` branch than to do some behind-the-scenes chat and tagging magic.

why are we dealing with backports and branches as if 3.7 was released? This is main branch development happening in zephyr right now, creating a 3.7 branch now and leaving 'zephyr' branch in a limbo is not really helpful. We will create a branch for 3.7 when the time comes, I do not think we need to worry about it now, this is mainstream development still.

**Comment by stephanosio** [comment]
> why are we dealing with backports and branches as if 3.7 was released? This is main branch development happening in zephyr right now, creating a 3.7 branch now and leaving 'zephyr' branch in a limbo is not really helpful. We will create a branch for 3.7 when the time comes, I do not think we need to worry about it now, this is mainstream development still.

Fair enough, using `zephyr` as the "trunk" branch and creating `zephyr-vx.y` release branches as they are needed in the future then, as loosely documented in https://docs.zephyrproject.org/latest/develop/modules.html#module-repositories.

**Comment by ithinuel** [comment]
> Fair enough, using zephyr as the "trunk" branch and creating zephyr-vx.y release branches as they are needed in the future then, as loosely documented in [docs.zephyrproject.org/latest/develop/modules.html#module-repositories](https://docs.zephyrproject.org/latest/develop/modules.html#module-repositories).

It’s worth noting that the current `zephyr` branch is an old head and the current `manifest-rev` pulled by zephyr is `main` and is not related to the old `zephyr`.
So the old `zephyr` head would need to be tagged with some name or its commits may be garbage collected.

**Review by ithinuel — APPROVED** [review]

**Comment by ithinuel** [comment]
Was merged by setting `zephyr`'s branch to this hash.

**Pull request closed without merging**